<a href="https://colab.research.google.com/github/PosgradoMNA/Proyecto-final-Equipo-135/blob/main/Semana_3_Actividad_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Parte 1: Fundamentos de bases de datos**





1.   Fundamentos de bases de datos y bases para ciencia de datos.

*Para mi, mi primer contacto fue con los documentos de Excel, como bien son citados los documentos tienen cierta base "estructurada" que los permite ser manejados de una manera mucho mas eficiente, pero en muchos casos no le quita lo complicado al buscar, filtrar, remplazar, realizar operaciones, etcetera...*

2.   Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.

*Conforme fui intentando desarrollar herramientas que me permitieran simplificar mi trabajo, descubri poco a poco la parte de la programación, para poder acomodar, eliminar errores (como los campos vacios al intentar capturar información) y muchas mejoras mas, pero despues descubrí MS Access, me ayudo a simplificar incluso mucho mas con las API que podrían desarrollarse para hacer mucho mas amigable e incluso eficiente la captura de información, poco despues comence a saber sobre estrucuras mucho mas "profesionales" como SQL, MySQL y Oracle aunque estas solo las llegue a utilizar para poder comprender la forma en la que trabajaban ya que fui entrenado para poder utilizar una herramienta que se llama Crystal Reports con la que comenzamos a hacer reportes para todos los departamentos de la compañía para la que trabajaba*

# **Parte 2: Selección y limpieza de los Datos en Python**

Para esta parte, utilizaremos un set de datos ubicado en la siguiente dirección:
https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv

*esta información será accedida a travez de Pandas (PD)
Para poder mandar a llamar la base de datos primero tenemos que establecer un path de tal forma que Python/Pandas sepa donde se encuentra el documento con el que estara trabajando.*

*y sera con las siguientes lineas a continuación:*


In [ ]:
import pandas as pd

data = pd.read_csv('https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv')
print(data)

          ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20 

In [ ]:
data.shape #nos explica como es la forma de nuestro Dataset

(30000, 25)

In [ ]:
data.isnull().values.any() #nos dice si existen valores vacios

True

In [ ]:
data.isna().values.any() #nos dice si existen valores NA

True

In [ ]:
data.isnull().any() #nos dice si existen valores vacios en cada columna

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

In [ ]:
data.dropna(inplace = True) # Asigna en los lugares vacios un "True" para ir rellenando los campos

In [ ]:
data # despliega el Dataset

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[29958 rows x 25 columns]

Como podemos ver, ha cambiado el valor de los 30,000 a 29,958

In [ ]:
data.isna().values.any() #muestra si hay valores con NA

False

In [ ]:
data.isnull().values.any() #Muestra si hay valores vacios o "Empty"

False

In [ ]:
data.sort_index(ascending=True) #realiza un sorteo d elos datos de forma acendente

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[29958 rows x 25 columns]

In [ ]:
data.head() #muestra unicamente los encabezados (las 5 primeras filas)

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

# **Parte 3: Preparación de los datos**




---


1.   ¿Qué datos considero mas importantes? 
**En realidad, podríamos considerar todos los datos pueden ser utiles en algun punto de nuestro analisis** *(a menos de que sean campos de texto o "strings" que no sean concretos, pero aun así en algun momento podríamos estar buscando palabras claves en estos campos por lo que, en general no podríamos descartar tan facilmente los campos)*
**Para este caso, los campos que consideramos que podrían ser importantes serían los siguientes: **

| Variables consideradas                    | Rason por la que se consideraron    | 
| ------------------------------- | ------------------------ |
| X1  = Amount of the given credit  | Tener los montos totales de credito otorgados  |
| X4  = Marital status  | Entender si cuenta con dependientes  |
| X5  = Age (year).   | Capacidad de pago / Trabajo  |
| X6  = the repayment status in Sep, 2005  | Desarrollo de estadisticas y status de pagos  |
| X7  = the repayment status in Aug, 2005  | 
| X8  = the repayment status in Jul, 2005  | 
| X9  = the repayment status in Jun, 2005  | 
| X10 = the repayment status in May, 2005  | 
| X11 = the repayment status in Apr, 2005  | 
| X12 = amount of bill statement in Sep, 2005  | Para el desarrollo de histogramas de pagos  |
| X13 = amount of bill statement in Aug, 2005  | 
| X14 = amount of bill statement in Jul, 2005  | 
| X15 = amount of bill statement in Jun, 2005  | 
| X16 = amount of bill statement in May, 2005  | 
| X17 = amount of bill statement in Apr, 2005  | 
| X18 = amount paid in September, 2005  | Para el desarrollo de histogramas y modelos de pago por montos  |
| X19 = amount paid in September, 2005  | 
| X20 = amount paid in September, 2005  | 
| X21 = amount paid in September, 2005  | 
| X22 = amount paid in September, 2005  | 
| X23 = amount paid in September, 2005  | 
 
 **Lista de variables no consideradas como "importantes**


| Variables excluidas       | Razón de exclusión    | 
| --- | --- |
| X2 = Gender | El genero solo podría aportar informacion estadística, no para la resolución de pagos  |
| X3 = Education  | Solo funcionaría con fines estadisticos ya que no esta directamente relacionada a la capacidad de pagos de creditos  |



2.   ¿Por qué?
**Para mi, lo mas importante es primero definir que información será la que se utilizara para poder partir con dicha información y el resto poder discriminarla** *(incluso por la optimización de los procesos y velocidad de manejo de información)*



3.   ¿Se eliminaron o reemplazaron datos nulos?
**Cuando estos pudieron ser remplazados, fueron remplazados, cuando no, fueron eliminados de tal forma que no contaminaran la información colectada**

4.   ¿Qué se hizo y por qué?

5.   ¿Es necesario ordenar los datos para el análisis? Sí / No / 
**No es necesario, pero se realizo por "estética" y para asegurar que toda la información estuviera organizada de forma Ascendente**
6.   ¿Por qué?

7.   ¿Existen problemas de formato que deban solucionar antes del proceso de 
modelado? Sí / No / 
**SI**
8.   Por qué.
**Por citar algunos ejemplos, podemos tener campos de fechas que no concuerden con lo que estaremos buscando, por ejemplo AAAA-MM-DD, si fue capturado como AAAA-DD-MM esto, en nuestro analizis nos traera inconvenientes y errores en el despliegue de información correcta**
9.   ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?
**Para este Dataset se realizaron revisiónes de los campos vaciós, se rellenaron con las medias de los datos para "estandarizarlo", y no perder información que podría ser confiable para la entrega de nuestros resultados.**